src/impl_owned_array.rs

```diff
@@ -59,14 +59,101 @@ impl<A> Array<A, Ix0>
 impl<A, D> Array<A, D>
 where D: Dimension
 {
+    /// Returns the offset (in units of `A`) from the start of the allocation
+    /// to the first element, or `None` if the array is empty.
+    fn offset_from_alloc_to_logical_ptr(&self) -> Option<usize>
+    {
+        if self.is_empty() {
+            return None;
```

Comment:
Does the differentiating the `None` case really buy anything if it is defaulted to zero at all call sites in any case?

Reply:
I've been puzzling over that, not sure

Reply:
> `(Vec<A>, Option<NonZeroUsize>)`

`NonZeroUsize` is nonsense, I was thinking of the length and not the index. But I still think either `Option<usize>` or `*mut A` would be preferable.

Reply:
Ok, why is option good? I think keeping this as option here is ok

Reply:
Because it avoids returning a value (zero) which cannot actually be used to index into the vector (which is empty) even tough it can be used to offset the pointer. Handling this case by a bit of additional wording in the doc comment is failure prone.

`Option<usize>` means calling code needs to acknowledge and handle this case (yielding `None` instead of an invalid index zero in the empty case). `*mut A` at least avoids this by always be useful as a valid pointer even if dangling.

Reply:
This is for into_raw_vec_with_offset right (the public method) and not for offset_from_alloc_to_logical_ptr? just checking

Reply:
Yes, as said in the beginning. I think keeping the `Option` here makes sense if it is also exposed at the public API. And it should be exposed at the public API, just replacing it by zero for the empty case is a foot gun because the invariants get unnecessarily complicated.

Reply:
I've changed it around to use `Option<usize>` again, it seems best.